contracts/JoeBarV2.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.6.12;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+interface IBarRewarder {
+    function claimReward() external;
+}
+
+// JoeBar is the coolest bar in town. You come in with some Joe, and leave with more! The longer you stay, the more Joe you get.
+//
+// This contract handles swapping to and from moJoe, JoeSwap's staking token.
+contract JoeBarV2 is Initializable, ERC20Upgradeable, OwnableUpgradeable {
+    using SafeMathUpgradeable for uint256;
+
+    IERC20Upgradeable public joe;
+    IBarRewarder public rewarder;
+    uint256 public entryFee;
+
+    event SetEntryFee(uint256 _entryFee);
+    event SetRewarder(address _rewarder);
+
+    // Define the Joe token contract
+    function initialize(IERC20Upgradeable _joe, uint256 _entryFee) public initializer {
+        __ERC20_init("JoeBarV2", "moJOE");
+        __Ownable_init();
+
+        joe = _joe;
+        setEntryFee(_entryFee);
+    }
+
+    // Enter the bar. Pay some JOEs. Earn some shares.
+    // Locks Joe and mints moJoe
+    function enter(uint256 _amount) external {
+        // Gets the rewards
+        rewarder.claimReward();
+
+        uint256 fee = _amount.mul(entryFee).div(10000);
+        uint256 amount = _amount.sub(fee);
```

Comment:
Let's rename `amount` to something else like `amountSubFee` otherwise it gets confused with `_amount`